lsp-headerline.el

```diff
@@ -130,128 +130,140 @@ narrow to the outer symbol."
-                  (concat
-                   (if lsp-headerline-breadcrumb-enable-symbol-numbers
-                       (concat
-                        (propertize (number-to-string index)
-                                    'face
-                                    'lsp-headerline-breadcrumb-symbols-face)
-                        " ")
-                     "")
-                   (if symbol2-icon
-                       (concat symbol2-icon symbol2-name)
-                     symbol2-name))))
-            (format "%s %s"
-                    (lsp-headerline--arrow-icon)
-                    (lsp-headerline--with-action symbol-to-append full-symbol-2))))
-        enumerated-symbols-hierarchy
-        " ")))))
+      (seq-map (-lambda (((symbol-to-append &as &DocumentSymbol :deprecated? :name)
+                          . index))
+                 (let* ((symbol2-name
+                         (propertize name
+                                     'font-lock-face
+                                     (if deprecated?
+                                         'lsp-headerline-breadcrumb-deprecated-face
+                                       'lsp-headerline-breadcrumb-symbols-face)))
+                        (symbol2-icon
+                         (lsp-headerline--symbol-icon symbol-to-append))
+                        (full-symbol-2
+                         (concat
+                          (if lsp-headerline-breadcrumb-enable-symbol-numbers
+                              (concat
+                               (propertize (number-to-string index)
+                                           'face
+                                           'lsp-headerline-breadcrumb-symbols-face)
+                               " ")
+                            "")
+                          (if symbol2-icon
+                              (concat symbol2-icon symbol2-name)
+                            symbol2-name))))
+                   (lsp-headerline--symbol-with-action symbol-to-append full-symbol-2)))
+               enumerated-symbols-hierarchy))))
 
 (defun lsp-headerline--build-string ()
   "Build the header-line string."
-  (seq-reduce (lambda (last-segment next-segment)
-                (concat last-segment
-                        (or (pcase next-segment
-                              ('project (lsp-headerline--build-project-string))
-                              ('file (lsp-headerline--build-file-string))
-                              ('path-up-to-project (or lsp-headerline--path-up-to-project-string
-                                                       (lsp-headerline--build-path-up-to-project-string)))
-                              ('symbols (lsp-headerline--build-symbols-string)))
-                            "")))
-              lsp-headerline-breadcrumb-segments ""))
+  (concat
+   (format "%s " (lsp-headerline--arrow-icon))
+   (s-join
+    (format " %s "
+            (lsp-headerline--arrow-icon))
+    (-flatten
```

Comment:
I am a bit concerned about the memory usage of this component, since more, and more the garbage collector performance will be hitting us. This code is on the hot path so we should try to optimize it as much as possible.

ATM we create intermittent lists as a result of the component call, -map and -flatten. Can we change each of the components result to "string" and then use mapconcat? Generally, headerline elements wont be that much but this operation will be called thousand times for a single session.

Reply:
I hadn't realized that this would gum up the GC.

OK, I can change the component results back to string, which is how it was before. There are some other simplifications from this commit that could stay, though. Should I make a commit to revert this one first, or should I continue at this commit?

Reply:
I don't see any problem in be implemented in a new PR

Reply:
no need to revert, it is not something urgent. We should also see how the code will look like when we implement https://github.com/emacs-lsp/lsp-mode/issues/1823#issuecomment-661312426 . IMO this will be very useful unique for emacs ecosystem feature.